doc/developer/logging.rst

```diff
@@ -239,3 +242,176 @@ Examples:
 * some field that is absolutely needed is :code:`NULL`
 * any other kind of data structure corruption that will cause the daemon
   to crash sooner or later, one way or another
+
+Thread-local buffering
+----------------------
+
+The core logging code in :file:`lib/zlog.c` allows setting up per-thread log
+message buffers in order to improve logging performance.  The following rules
+apply for this buffering:
+
+* Only messages of priority *DEBUG* or *INFO* are buffered.
+* Any higher-priority message causes the entire buffer to be flushed, thus
+  message ordering is preserved properly.
+* The buffers are only ever accessed by the thread they are created by.  This
+  means no locking is necessary.
```

Comment:
Just to clarify -- if I have multiple threads logging to separate buffers, and one thread pushes a higher priority log message, will the buffers for "all" threads be flushed, or only for the one thread which pushed the higher priority log message? If only the single thread is flushed, it seems like the message ordering is preserved per thread, but not between threads, correct?

Reply:
Indeed, the flush is only local and thus only ordering for that particular thread is preserved.

The way I look at it is, ordering of log messages from different threads is arbitrary either way unless the code manually imposes some locks, and if it does that it might as well call `zlog_tls_buffer_flush()`.

That said, I should probably add this to the docs :smile: